src/tls/server.c

```diff
@@ -552,45 +553,53 @@ server_init (const char *ws_path,
     }
 
   /* systemd socket activated? */
-  listen_fds = secure_getenv ("LISTEN_FDS");
-  if (listen_fds && check_sd_listen_pid ())
+  env_listen_fds = secure_getenv ("LISTEN_FDS");
+  if (env_listen_fds && check_sd_listen_pid ())
     {
-      if (strcmp (listen_fds, "1") != 0)
-        errx (1, "This program can only accept exactly one socket from systemd, but got passed %s", listen_fds);
-      server.listen_fd = SD_LISTEN_FDS_START;
-      debug ("Server ready. Listening to systemd activated socket fd %i", server.listen_fd);
+      char *endptr = NULL;
+      long n = strtol (env_listen_fds, &endptr, 10);
+      if (n < 1 || n > MAX_LISTEN_FDS || !endptr || *endptr != '\0')
+        errx (1, "Invalid $LISTEN_FDS value '%s'; this program supports up to 10 fds", env_listen_fds);
+      for (int i = 0, fd = SD_LISTEN_FDS_START; i < n; ++i, ++fd)
+        {
+          server.listen_fds[i] = fd;
```

Comment:
This is deeply weird.  We have an array which we fill with completely deterministic values.

I can almost imagine an optimisation where we rather record the number of fds and then any time we see an fd ready to poll which has a value between 3 and 3+num then we know it's a listener.  That's also deeply weird, though.  Also: it doesn't give us an otherwise-useless location to use as epoll's data ptr.

Reply:
Right, it's by and large an encoding of the fd number to be used with data.ptr, as we want to use pointers consistently. When we previously had the intermixed "data.fd for listen and data.ptr for connections", this was also error prone/weird. Hopefully this will go away with your threading rework anyway. Right now this still strikes me as the less bad approach, but I'm happy to change  this around if you have a cleaner idea.